driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/gridfs/CollectionTimeoutHelper.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal.gridfs;
+
+import com.mongodb.MongoExecutionTimeoutException;
+import com.mongodb.internal.time.Timeout;
+import com.mongodb.lang.Nullable;
+import com.mongodb.reactivestreams.client.MongoCollection;
+import reactor.core.publisher.Mono;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+final class CollectionTimeoutHelper {
+    private CollectionTimeoutHelper(){
+        //NOP
+    }
+    private static <T> MongoCollection<T> applyTimeout(final MongoCollection<T> collection,
+                                                               @Nullable final Timeout timeout) {
+        if (timeout != null && !timeout.isInfinite()) {
+            long remainingMs = timeout.remaining(MILLISECONDS);
+            if (remainingMs <= 0) {
+                throw new MongoExecutionTimeoutException("GridFS timed out");
+            }
+            return collection.withTimeout(remainingMs, MILLISECONDS);
+        }
+        return collection;
+    }
+
+
+    public static <T> MongoCollection<T> withNullableTimeout(final MongoCollection<T> collection,
+                                                                  @Nullable final Timeout timeout) {
+        return applyTimeout(collection, timeout);
+    }
+
+    public static <T> Mono<MongoCollection<T>> withNullableTimeoutMono(final MongoCollection<T> collection,
```

Comment:
We had a verbal discussion with @rozza , and I'm summarizing the outcome here for clarity:

When we construct a stream pipeline, we use Mono to encapsulate the collection with an associated timeout. Once the collection is wrapped in this Mono, the timeout is fixed and remains unchanged until the operation starts execution.

The asynchronous nature of reactive streams, especially in I/O-bound scenarios, makes it difficult to predict the exact timing of execution. Therefore, we employ Deferred Mono so that wrapping a collection with the remaining timeout happens at a time closer to the operation’s execution.

As for the use of a 'hot' Mono, there are instances where the entire code block is already wrapped within Mono.defer. In these situations, creating an additional deferred Mono for the collection is redundant. Hence, we have a method that returns a non-deferred Mono.